param.go

```diff
@@ -326,3 +342,52 @@ func (po paramObject) Build(c *Container) (reflect.Value, error) {
 	}
 	return dest, nil
 }
+
+// paramGroupedSlice is a param which produces a slice of values with the same
+// group name.
+type paramGroupedSlice struct {
+	// Name of the group as specified in the `group:".."` tag.
+	Group string
+
+	// Type of the slice.
+	Type reflect.Type
+}
+
+// newParamGroupedSlice builds a paramGroupedSlice from the provided type with
+// the given name.
+//
+// The type MUST be a slice type.
+func newParamGroupedSlice(group string, t reflect.Type) paramGroupedSlice {
+	return paramGroupedSlice{Group: group, Type: t}
+}
+
+func (pt paramGroupedSlice) Build(c *Container) (reflect.Value, error) {
+	k := key{group: pt.Group, t: pt.Type.Elem()}
+
+	for _, n := range c.producers[k] {
+		if err := n.Call(c); err != nil {
+			// TODO: better message
+			return _noValue, errWrapf(err, "failed to build %v", k)
+		}
+	}
+
+	items := c.groups[k]
+
+	// shuffle the list so users don't rely on the ordering of grouped values
+	items = shuffledCopy(items)
+
+	result := reflect.MakeSlice(pt.Type, len(items), len(items))
+	for i, v := range items {
+		result.Index(i).Set(v)
+	}
+	return result, nil
+}
+
+func shuffledCopy(items []reflect.Value) []reflect.Value {
+	newItems := make([]reflect.Value, len(items))
+	// TODO(abg): Inject rand to get some determinism during tests.
```

Comment:
good idea!

Reply:
We can get randomness implicitly if we track the values internally with a map[T]struct{}, translating to and from slices at the edges.

Reply:
Thinking out loud - should we favor explicitness in this case, since it’s all transparent to the anyways?